### DISCUSSION TITLE
Can Multiple Functions Use the Same .jar in Apache Pulsar?

### DISCUSSION THREAD

**Discussion by NurramoX** [apache/pulsar#22737] [Q&A] [answered]
Hi everyone,

I'm working with Apache Pulsar Functions and I have a question regarding .jar file usage:

Can multiple functions make use of the same .jar file that I uploaded? Specifically, what happens if I upload the same .jar file twice during the function registration process for two different functions?

I'm interested in understanding if there's any impact on performance, redundancy, or any potential conflicts.

Thanks in advance!

**Comment by lhotari** [comment, marked as answer]
You can have multiple functions in the same .jar file, but you would have to upload the file separately for each function.
It would be a useful feature to be able to define and run multiple functions together, however that feature doesn't exist.

Pulsar Functions does have the package manager feature ([PIP-50](https://github.com/apache/pulsar/wiki/PIP-50:-Package-Management)) which allows sharing the uploaded file, but that's doesn't cover creating, updating and running multiple functions together.

**Comment by NurramoX** [comment]
Thanks for the reply!

Just to clarify, am I understanding correctly that with the package manager, we can upload the .jar file only once, but we still need to deploy and run each function individually by making separate requests for each function deployment?

**Reply by lhotari** [reply]
exactly.
